src/app/api-server/src/flightsql_server.rs

```diff
@@ -0,0 +1,100 @@
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+/////////////////////////////////////////////////////////////////////////////////////////
+
+use std::net::SocketAddr;
+use std::sync::Arc;
+
+use arrow_flight::flight_service_server::FlightServiceServer;
+use datafusion::prelude::SessionContext;
+use futures::Future;
+use kamu::domain::CurrentAccountSubject;
+use kamu_adapter_flight_sql::{KamuFlightSqlService, SessionFactory, Token};
+use tokio::net::TcpListener;
+use tonic::transport::Server;
+use tonic::Status;
+
+/////////////////////////////////////////////////////////////////////////////////////////
+
+pub(crate) struct FlightSqlServer {
+    service: KamuFlightSqlService,
+    listener: TcpListener,
+}
+
+/////////////////////////////////////////////////////////////////////////////////////////
+
+impl FlightSqlServer {
+    pub async fn new(
+        address: std::net::IpAddr,
+        port: Option<u16>,
+        base_catalog: dill::Catalog,
+    ) -> Self {
+        let service = kamu_adapter_flight_sql::KamuFlightSqlService::builder()
+            .with_server_name(crate::BINARY_NAME, crate::VERSION)
+            .with_session_factory(Arc::new(SessionFactoryImpl { base_catalog }))
+            .build();
+
+        let listener = TcpListener::bind((address, port.unwrap_or_default()))
+            .await
+            .unwrap();
+
+        Self { service, listener }
+    }
+
+    pub fn local_addr(&self) -> SocketAddr {
+        self.listener.local_addr().unwrap()
+    }
+
+    pub fn run(self) -> impl Future<Output = Result<(), impl std::error::Error>> {
+        Server::builder()
+            .add_service(FlightServiceServer::new(self.service))
+            .serve_with_incoming(tokio_stream::wrappers::TcpListenerStream::new(
+                self.listener,
+            ))
+    }
+}
+
+/////////////////////////////////////////////////////////////////////////////////////////
+
+struct SessionFactoryImpl {
+    base_catalog: dill::Catalog,
+}
+
+#[async_trait::async_trait]
+impl SessionFactory for SessionFactoryImpl {
+    #[tracing::instrument(level = "debug", skip_all, fields(username))]
+    async fn authenticate(&self, username: &str, password: &str) -> Result<Token, Status> {
+        // TODO: SEC: Real auth via app token
+        if username == "kamu" && password == "kamu" {
+            Ok(String::new())
+        } else {
+            Err(Status::unauthenticated("Invalid credentials!"))
+        }
+    }
+
+    #[tracing::instrument(level = "debug", skip_all)]
+    async fn get_context(&self, _token: &Token) -> Result<Arc<SessionContext>, Status> {
+        let subject = CurrentAccountSubject::Anonymous(
+            kamu::domain::AnonymousAccountReason::NoAuthenticationProvided,
+        );
```

Comment:
Doesn't this thing go through our HTTP authentication middleware? If no, maybe it should.

Reply:
Arrow Flight is built on top of gRPC, which uses HTTP/2 as transport, so there is a chance we can reuse auth middleware. But I'm still in "just make it work" mode here.